moly-kit/src/widgets/messages.rs

```diff
@@ -245,12 +254,19 @@ impl Messages {
                         .map(|b| (b.name.as_str(), Some(b.avatar.clone())))
                         .unwrap_or(("Unknown bot", Some(Picture::Grapheme("B".into()))));
 
-                    // If the message is empty, display a loading animation.
-                    let item = if message.is_writing && message.body.is_empty() {
+                    // If the message is empty and still writing, display a loading animation
+                    let body_text = message.visible_text();
+                    let item = if message.is_writing && body_text.is_empty() && !message.has_stages() {
                         let item = list.item(cx, index, live_id!(LoadingLine));
-
                         item.message_loading(id!(text.loading)).animate(cx);
-
+                        item
+                    } else if message.has_stages() {
+                        // Use specialized DeepInquireBotLine for messages with stages
+                        let item = list.item(cx, index, live_id!(DeepInquireBotLine));
+                        
+                        // Update the DeepInquireBotLine with the message content
+                        item.as_deep_inquire_bot_line().set_message(cx, message, message.is_writing);
+                        
```

Comment:
I didn't have much time for it now, but here's where I think we should have an API that allows MolyKit to defer/delegate the message rendering to a BotLine widget, that uses the base one by default, but app devs can extend.
So that you can create your own very custom client, that uses it's own formatting, and then Messages widget says "ok this client maps to this line renderer".
I know there's an option to make everything more dynamic through text (similar to thinking blocks) but I think there are great benefits to having a type safe implementation for providers/messages that are complex.
**tl;dr:** we should think about this, and make it in a way that DeepInquire serves as an example on how to add other very custom clients.